lib/resty/rediscluster.lua

```diff
@@ -171,30 +192,68 @@ local function try_hosts_slots(self, serv_list)
                 -- this can differ from self.config.serv_list if a cluster is resized (added/removed nodes)
                 local servers = { serv_list = {} }
                 for n = 1, #slots_info do
+                    -- Redis 6: { 0, 5460, { "172.26.0.21", 6379, "b050a28d860201c487891ecde77efc8a3935ea51" }, { "172.26.0.25", 6379, "20d96bc4a640fd244699e29274b2c94962def80e" } }
+                    -- Redis 7: { 0, 5460, { "rc-node-1", 6379, "3a4f2814802ca25a5c92dfc167674802126557d3", { "ip", "172.26.0.11" } }, { "rc-node-5", 6379, "f14b2cf710613f517374aefe7b84d9f013102ebe", { "ip", "172.26.0.15" } } }
+                    -- Redis 7: { 0, 5460, { "172.26.0.11", 6379, "64ff3bc4e8f4587c92bcfaa6c085e67417d44d5d", {} }, { "172.26.0.15", 6379, "9a63c473b928203b141d20a46be6784eb320ba0c", {} } }
                     local sub_info = slots_info[n]
-                    --slot info item 1 and 2 are the subrange start end slots
+                    -- slot info item 1 and 2 are the subrange start end slots
                     local start_slot, end_slot = sub_info[1], sub_info[2]
 
-                    -- generate new list of servers
+                    -- generate new list of servers, including IPs and hostnames
+                    local sub_info_j
+                    local sub_info_j_ip
                     for j = 3, #sub_info do
-                      servers.serv_list[#servers.serv_list + 1 ] = { ip = sub_info[j][1], port = sub_info[j][2] }
+                        -- Redis 6: { "172.26.0.21", 6379, "b050a28d860201c487891ecde77efc8a3935ea51" }
+                        -- Redis 7: { "rc-node-1", 6379, "3a4f2814802ca25a5c92dfc167674802126557d3", { "ip", "172.26.0.11" } }
+                        -- Redis 7: { "172.26.0.11", 6379, "64ff3bc4e8f4587c92bcfaa6c085e67417d44d5d", {} }
+                        sub_info_j = sub_info[j]
+                        -- Redis 6: nil
+                        -- Redis 7: { "ip", "172.26.0.11" }
+                        -- Redis 7: {}
+                        sub_info_j_ip = sub_info_j[4]
+                        if sub_info_j_ip and type(sub_info_j_ip) == "table" and
+                           next(sub_info_j_ip) and sub_info_j_ip[2] ~= sub_info_j[1]
+                        then
+                            redis7_with_host = true
+                            -- prioritize IP address over hostname as IP does not suffer from stale DNS cache
+                            servers.serv_list[#servers.serv_list + 1] = { ip = sub_info_j_ip[2], port = sub_info_j[2] }
+                        else
+                            redis7_with_host = false
+                        end
```

Comment:
To me, this is still very unstructured and error-prone code. I think we can do better even if this means more refactoring

Reply:
Will refactor this part as this part is related to the bug.